### PR TITLE
Thread-safe locking

### PR DESCRIPTION
**Pull request by aspiers**
Make the library thread-safe, so that multiple clients (web UI or `crowbarctl`) can simultaneously read/write upgrade state in a safe manner.

Still TODO:

- [ ] fix pending test
- [ ] change `#start_step` and `#end_step` to take a step name as mandatory parameter, and add (and test) checks within the critical code regions protected by locks which validate that it's permitted to start or end the given step

### CODE REVIEW

File: crowbar_framework/lib/crowbar/upgrade_status.rb [hunk 1]
```diff
@@ -97,7 +97,7 @@ def finished?
 
     def save
       progress_file_path.open("w") do |f|
-        f.write(JSON.pretty_generate(progress))
+        f.write(YAML.dump(progress))
```
Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: still not convinced why to use yaml for the state, at least not from the commit message. isn't readability a better reason to use yaml, not some internal restrictions with symbols?
and not that it matters for much for crowbar, but when we have to read the file from our automation (bash) wouldn't json be better?

Comment (jsuchome) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: Shoudln't we rather go back to symbolizing the string values, like we've had before in https://github.com/crowbar/crowbar-core/commit/b7909220d9f5572b4f695d1b81f093143e72cb8f (and reverted by https://github.com/crowbar/crowbar-core/commit/52589eee5b121fa198570967cb2bcfe830feb0e8)

Comment (aspiers) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: @MaximilianMeister Yes, sorry for the bad commit message - I was very tired when working on this and forgot to put justification. Will fix that!

@jsuchome No I don't think we should, otherwise I wouldn't have proposed this change to YAML.

**Comment by aspiers** [comment]
Commit messages improved.  Thanks for keeping me honest on this, and please point out any time in future where I forget to write clear commit messages ;-)

File: crowbar_framework/lib/crowbar/upgrade_status.rb (hunk 1 above)
Comment (jsuchome) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: Shouldn't it be in separate PR then? It's not related to thread locking and current library does not work with its mix of strings and symbols.

Comment (aspiers) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: No, it's related because it was necessary to fix this in order to get the test suite working.

Comment (jsuchome) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: Test suite might be working until now, but library is not, because at some point it checks for strings instead of symbols or vice versa. It is fixed by either deep_symbolize_keys call or by moving to YAML, but it is not specific to thread safe topic

**Comment by MaximilianMeister** [comment]
@aspiers thanks for clarification, but TBH i am still not convinced with using yaml, and it feels to me we are only using it because of the serialization feature.

the question is also: why does the status file even need to be human readable?

File: crowbar_framework/lib/crowbar/upgrade_status.rb [hunk 2]
```diff
@@ -1,27 +1,42 @@
 module Crowbar
   class UpgradeStatus
+    attr_reader :progress_file_path
     attr_accessor :progress
 
     # Return the current state of upgrade process.
     # We're keeping the information in the file so is accessible by
     # external applications and different crowbar versions.
-    def initialize
+    def initialize(json_file = "/var/lib/crowbar/upgrade/progress.json")
```
Comment (jsuchome) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: `yaml_file = "/var/lib/crowbar/upgrade/progress.yml`

**Comment by jsuchome** [comment]
[(5) S50P2 (S49P4) Implement a state machine for the upgrade](https://trello.com/c/OWehKX6K/126-5-s50p2-s49p4-implement-a-state-machine-for-the-upgrade)

File: crowbar_framework/lib/crowbar/upgrade_status.rb (hunk 2 above)
Comment (aspiers) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: Thanks! Will fix.

**Comment by aspiers** [comment]
> @aspiers thanks for clarification, but TBH i am still not convinced with using yaml

Why not? And why do you think JSON would be better? Let's discuss facts not feelings ;-)

> and it feels to me we are only using it because of the serialization feature.

Not only, but avoiding the need to carefully convert certain parts of the data structure between string and symbol on read/write is certainly one advantage.

> the question is also: why does the status file even need to be human readable?

Because it makes debugging way easier!  One simple example:

    watch -d cat /var/lib/crowbar/upgrade/progress.yml

File: crowbar_framework/lib/crowbar/upgrade_status.rb (hunk 1 above)
Comment (jsuchome) [inline crowbar_framework/lib/crowbar/upgrade_status.rb]: I extracted the commit here https://github.com/crowbar/crowbar-core/pull/769. I believe we need this before the thread safe locking is done to finally make use of the progress library and unblock UI and QA.

**Comment by aspiers** [comment]
Fixed and rebased.
